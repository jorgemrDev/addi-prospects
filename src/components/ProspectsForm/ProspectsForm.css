label {
  font-weight: bold;
}

.btn {
  display: inline-block;
  margin-bottom: 0;
  font-weight: 400;
  text-align: center;
  vertical-align: middle;
  -ms-touch-action: manipulation;
  touch-action: manipulation;
  cursor: pointer;
  background-image: none;
  border: 1px solid transparent;
  white-space: nowrap;
  padding: 6px 12px;
  font-size: 14px;
  line-height: 1.42857143;
  border-radius: 3px;
  -webkit-user-select: none;
  -moz-user-select: none;
  -ms-user-select: none;
  user-select: none;
}

.btn-primary {
  color: #fff;
  background-color: #364147;
}

h1 {
  color: #fff;
  background-color: #364147;
}

.success-info {
  margin-top: 70px;
  margin-left: 200px;
}
