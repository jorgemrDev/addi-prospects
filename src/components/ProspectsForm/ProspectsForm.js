import React, { useState } from "react";
import { getPerson } from "../../services/national-registry/nationalRegistryService";
import { getJudicialRecords } from "../../services/judicialRecords/judicialRecordsService";
import { getProspectQualification } from "../../services/prospectQualification/prospectQualificationService";
import { putToProspect } from "../../services/crm/crmService";
import "./ProspectsForm.css";

const STATUS = {
  IDLE: "IDLE",
  SUBMITTED: "SUBMITTED",
  SUBMITTING: "SUBMITTING",
  COMPLETED: "COMPLETED",
};

export default function ProspectsForm({ prospect }) {
  const [status, setStatus] = useState(STATUS.IDLE);
  const [errors, setErrors] = useState({});
  const [validations, setValidations] = useState({});
  const isValid =
    Object.keys(errors).length === 0 && Object.keys(validations).length === 0;

  function getErrors(nationalRegistryData) {
    const result = {};
    if (nationalRegistryData.nationalIdNumber !== prospect.nationalIdNumber)
      result.nationalIdNumber =
        "National Identification is different from the National Registry";
    if (nationalRegistryData.birthdate !== prospect.birthdate)
      result.birthdate = "Birthdate different from the National Registry";
    return result;
  }

  async function convertToProspect() {
    await putToProspect(parseInt(prospect.nationalIdNumber));
  }

  async function handleSubmit(event) {
    event.preventDefault();
    setStatus(STATUS.SUBMITTING);
    //exectue two first validations on paralell
    const [nationalRegistryInfo, judicialRecords] = await Promise.all([
      getPerson(parseInt(prospect.nationalIdNumber)),
      getJudicialRecords(parseInt(prospect.nationalIdNumber)),
    ]);

    let errors = getErrors(nationalRegistryInfo[0]);
    let validations = {};
    if (judicialRecords && judicialRecords[0] && judicialRecords[0].records) {
      validations.judicialRecords = "Have judicial records";
    }

    let prospectQualification = await getProspectQualification(
      nationalRegistryInfo[0]
    );

    if (prospectQualification && prospectQualification <= 60) {
      validations.prospectQualification = `The prospect qualification is not greater 60 points as required. Is: ${prospectQualification}`;
    }

    if (Object.keys(errors).length > 0 || Object.keys(validations).length > 0) {
      setErrors(errors);
      setValidations(validations);
      setStatus(STATUS.SUBMITTED);
    } else {
      await convertToProspect();
      setStatus(STATUS.COMPLETED);
    }
  }

  if (status === STATUS.COMPLETED) {
    return <h1>Succesfully converted to Prospect!</h1>;
  }

  return (
    <>
      <h1>Prospect Info</h1>
      {!isValid && (
        <div role="alert">
          <p>The lead cant not be converted to prospect due to:</p>
          <ul>
            {Object.keys(validations).map((key) => {
              return <li key={key}>{validations[key]}</li>;
            })}
          </ul>
        </div>
      )}
      <form onSubmit={handleSubmit}>
        <div>
          <label htmlFor="idNationalNumber">Id National Number</label>
          <p htmlFor="idNationalNumber"> {prospect.nationalIdNumber} </p>
          <br />
          <p role="alert">
            {status === STATUS.SUBMITTED && errors.nationalIdNumber}
          </p>
        </div>

        <div>
          <label htmlFor="idNationalNumber">First Name</label>
          <p htmlFor="idNationalNumber"> {prospect.firstName} </p>
          <br />
        </div>

        <div>
          <label htmlFor="idNationalNumber">Last Name</label>
          <p htmlFor="idNationalNumber"> {prospect.lastName} </p>
          <br />
        </div>

        <div>
          <label htmlFor="idNationalNumber">Birthdate</label>
          <p htmlFor="idNationalNumber"> {prospect.birthdate} </p>
          <br />
          <p role="alert">{errors.birthdate}</p>
        </div>

        <div>
          <label htmlFor="idNationalNumber">Email</label>
          <p htmlFor="idNationalNumber"> {prospect.email} </p>
          <br />
        </div>

        <div>
          <input
            type="submit"
            className="btn btn-primary"
            value="Convert to Prospect"
            disabled={status === STATUS.SUBMITTING}
          />
        </div>
      </form>
    </>
  );
}
