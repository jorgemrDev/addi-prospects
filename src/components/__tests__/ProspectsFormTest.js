import React from "react";
import ProspectsForm from "../ProspectsForm";
import renderer from "react-test-renderer";

const emptyLead = {
  idNationalNumber: "",
};

test("Render ProspectsForm correctly", () => {
  const component = renderer.create(
    <ProspectsForm prospect={emptyLead}></ProspectsForm>
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
