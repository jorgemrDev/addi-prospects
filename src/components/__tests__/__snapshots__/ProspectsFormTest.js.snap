// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render ProspectsForm correctly 1`] = `
Array [
  <h1>
    Prospect Info
  </h1>,
  <form
    onSubmit={[Function]}
  >
    <div>
      <label
        htmlFor="national-id-number"
      >
        Id National Number
      </label>
      <p
        id="national-id-number"
      >
         
         
      </p>
      <br />
      <p
        role="alert"
      />
    </div>
    <div>
      <label
        htmlFor="first-name"
      >
        First Name
      </label>
      <p
        htmlFor="first-name"
      >
         
         
      </p>
      <br />
    </div>
    <div>
      <label
        htmlFor="last-name"
      >
        Last Name
      </label>
      <p
        htmlFor="last-name"
      >
         
         
      </p>
      <br />
    </div>
    <div>
      <label
        htmlFor="birthdate"
      >
        Birthdate
      </label>
      <p
        htmlFor="birthdate"
      >
         
         
      </p>
      <br />
      <p
        role="alert"
      />
    </div>
    <div>
      <label
        htmlFor="email"
      >
        Email
      </label>
      <p
        htmlFor="email"
      >
         
         
      </p>
      <br />
    </div>
    <div>
      <input
        className="btn btn-primary"
        disabled={false}
        type="submit"
        value="Convert to Prospect"
      />
    </div>
  </form>,
]
`;
